src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/TypesThatOwnDisposableFieldsShouldBeDisposable.cs

```diff
@@ -42,83 +42,78 @@ public override void Initialize(AnalysisContext context)
+                return;
             }
 
-            public void AnalyzeSymbol(SymbolAnalysisContext symbolContext)
+            var disposableFields = namedType
+                .GetMembers()
+                .OfType<IFieldSymbol>()
+                .Where(f => !f.IsStatic && disposeAnalysisHelper.IsDisposable(f.Type))
+                .ToSet();
+            if (disposableFields.Count == 0)
             {
-                INamedTypeSymbol namedType = (INamedTypeSymbol)symbolContext.Symbol;
-                if (_disposeAnalysisHelper.IsDisposable(namedType))
-                {
-                    return;
-                }
+                return;
+            }
 
-                IEnumerable<IFieldSymbol> disposableFields = from member in namedType.GetMembers()
-                                                             where member.Kind == SymbolKind.Field && !member.IsStatic
-                                                             let field = member as IFieldSymbol
-                                                             where _disposeAnalysisHelper.IsDisposable(field.Type)
-                                                             select field;
-                if (!disposableFields.Any())
-                {
-                    return;
-                }
+            var disposableFieldNamesBuilder = ImmutableArray.CreateBuilder<string>();
 
-                var disposableFieldsHashSet = new HashSet<ISymbol>(disposableFields);
-                IEnumerable<TTypeDeclarationSyntax> classDecls = GetClassDeclarationNodes(namedType, symbolContext.CancellationToken);
-                foreach (TTypeDeclarationSyntax classDecl in classDecls)
+            ctx.RegisterOperationAction(context => AnalyzeOperation(context, namedType, disposableFields, disposableFieldNamesBuilder), OperationKind.SimpleAssignment, OperationKind.FieldInitializer);
+            ctx.RegisterSymbolEndAction(context => AnalyzeSymbolEnd(context, disposableFieldNamesBuilder.ToImmutable()));
+        }
+
+        private static void AnalyzeOperation(OperationAnalysisContext ctx, INamedTypeSymbol parent, ISet<IFieldSymbol> disposableFields, ImmutableArray<string>.Builder disposableFieldNamesBuilder)
+        {
+            if (ctx.Operation is IAssignmentOperation { Target: IFieldReferenceOperation field } assignment
+                && IsObjectCreation(assignment.Value)
+                && !ctx.Options.IsConfiguredToSkipAnalysis(Rule, field.Field.Type, parent, ctx.Compilation)
+                && disposableFields.Contains(field.Field))
+            {
+                disposableFieldNamesBuilder.Add(field.Field.Name);
+            }
+            else if (ctx.Operation is IFieldInitializerOperation initializer && IsObjectCreation(initializer.Value))
+            {
+                var candidateFields = initializer.InitializedFields
+                    .Where(f => !ctx.Options.IsConfiguredToSkipAnalysis(Rule, f.Type, parent, ctx.Compilation))
```

Comment:
Same here. Move this call inside the foreach loop OR consider adding a local function that has the `disposableFieldNamesBuilder.Add` guarded by `!ctx.Options.IsConfiguredToSkipAnalysis(...)` check.

Reply:
Alternatively, you can also consider switching `disposableFieldNamesBuilder` to `disposableFieldsBuilder` and move the `IsConfiguredToSkipAnalysis` check in the SymbolEnd action just prior to reporting the diagnostic. Any of these approaches is fine.